### PR TITLE
ISPN-11599 Out of the box testing with Infinispan Server and JUnit 5 

### PR DESCRIPTION
**Pull request by karesti**
Create an out of the box testing utility for JUnit 5
https://issues.redhat.com/browse/ISPN-11599

* The JUnit 5 commit is rebased from another PR 
https://issues.redhat.com/browse/ISPN-11444

* Provide an API to allow the user to create a standalone embedded server without forcing them to provide a xml file
https://issues.redhat.com/browse/ISPN-11505

* JUnit5 Infinispan Server Extension
https://issues.redhat.com/browse/ISPN-11600

### CODE REVIEW

**Comment by karesti** [comment]
@danberindei I updated to use the junit 5 with jupiter and jupiter vintage. not sure if everything is working as it should

File: server/testdriver/core/pom.xml [hunk 1]
```diff
@@ -68,5 +68,33 @@
...
+            <artifactId>maven-resources-plugin</artifactId>
+            <executions>
+               <execution>
+                  <id>copy-server-config</id>
+                  <phase>validate</phase>
+                  <goals>
+                     <goal>copy-resources</goal>
+                  </goals>
+                  <configuration>
+                     <outputDirectory>${project.basedir}/src/main/resources</outputDirectory>
```
Comment (danberindei) [inline server/testdriver/core/pom.xml:84]: Please copy it to `target/generated-sources/`, so you don't have to add it to `.gitignore`

@tristantarrant isn't there a way to use the default server configuration from `server/runtime` without copying it?

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java [hunk 2]
```diff
@@ -16,29 +31,21 @@
...
-import org.wildfly.security.x500.cert.X509CertificateBuilder;
-
 /**
  * @author Tristan Tarrant &lt;tristan@infinispan.org&gt;
  * @since 10.0
  **/
 public abstract class AbstractInfinispanServerDriver implements InfinispanServerDriver {
+   public static final String DEFAULT_CLUSTERED_INFINISPAN_CONFIG_FILE_NAME = "infinispan.xml";
+
+   private static final List<String> DEFAULT_INFINISPAN_CONFIG_FILES = Arrays.asList(
```
Comment (danberindei) [inline server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java:46]:
```suggestion
   private static final List<String> DEFAULT_INFINISPAN_CONFIG_FILES = 
         Arrays.asList(DEFAULT_CLUSTERED_INFINISPAN_CONFIG_FILE_NAME);
```

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java [hunk 3]
```diff
@@ -84,22 +91,32 @@ public void prepare(String name) {
          throw new RuntimeException("Failed to create server configuration directory " + confDir);
       }
       URL configurationFileURL;
-      if (new File(configuration.configurationFile()).isAbsolute()) {
+      if (DEFAULT_INFINISPAN_CONFIG_FILES.contains(configuration.configurationFile())) {
```
Comment (danberindei) [inline server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java:94]: Is this change really necessary?

"infinispan.xml" is not an absolute path, so it would use the `else` branch, which looks exactly the same.

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java [hunk 4]
```diff
@@ -84,22 +91,32 @@ public void prepare(String name) {
...
       if (configurationFileURL == null) {
          throw new RuntimeException("Cannot find test configuration file: " + configuration.configurationFile());
       }
-      Path configurationFilePath;
+
       try {
-         configurationFilePath = Paths.get(configurationFileURL.toURI());
-         // Recursively copy the contents of the directory containing the configuration file to the test target
-         Util.recursiveDirectoryCopy(configurationFilePath.getParent(), confDir.toPath());
+         if (DEFAULT_INFINISPAN_CONFIG_FILES.contains(configuration.configurationFile())) {
```
Comment (danberindei) [inline server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java:106]: Here you're only copying the configuration file. What happens with the other files the server needs, like `log4j2.xml`?

File: commons-test/pom.xml [hunk 5]
```diff
@@ -25,8 +25,13 @@
...
       <dependency>
-         <groupId>junit</groupId>
-         <artifactId>junit</artifactId>
+         <groupId>org.junit.jupiter</groupId>
+         <artifactId>junit-jupiter-engine</artifactId>
+         <scope>provided</scope>
+      </dependency>
+      <dependency>
+         <groupId>org.junit.vintage</groupId>
+         <artifactId>junit-vintage-engine</artifactId>
```
Comment (danberindei) [inline commons-test/pom.xml:34]: I don't think adding the dependencies here is enough, you also have to add them to the surefire plugin instead of the `surefire-junit47` dependency:

https://junit.org/junit5/docs/current/user-guide/#running-tests-build-maven

File: server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/InfinispanServerRuleBuilder.java [hunk 6]
```diff
@@ -24,6 +25,18 @@
    private boolean jmx;
    private boolean parallelStartup = true;
 
+   /**
+    * Use this method to instantiate a single clustered embedded server
+    *
+    * @return InfinispanServerRule
+    */
+   public static InfinispanServerRule standalone() {
+      return new InfinispanServerRuleBuilder(DEFAULT_CLUSTERED_INFINISPAN_CONFIG_FILE_NAME)
```
Comment (danberindei) [inline server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/InfinispanServerRuleBuilder.java:34]: IMO we should add a constructor without any parameters that uses the default server configuration.

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java (hunk 3 above)
Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java:94]: it raises an exception

Comment (danberindei) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java:94]: Are you sure? The code looks exactly the same to me.

File: server/testdriver/core/pom.xml (hunk 1 above)
Comment (tristantarrant) [reply server/testdriver/core/pom.xml:84]: We don't want to carry around any version of the server configuration in the testdriver-core module. CONTAINER mode would use the default configs present in the image and EMBEDDED mode should only be used by our own integration tests

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java (hunk 2 above)
Comment (tristantarrant) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java:46]: I would not have a default configuration name specified. In CONTAINER mode we can simply start the server with no configuration and it will use the default `infinispan.xml`

File: server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/InfinispanServerRuleBuilder.java (hunk 6 above)
Comment (tristantarrant) [reply server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/InfinispanServerRuleBuilder.java:34]: s/standalone()/server()/

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java (hunk 3 above)
Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java:94]: new File(configuration.configurationFile()) -> raises exception for the file when it is in the jar

Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java:94]: but the code can be refactored for sure to avoid duplication

Comment (danberindei) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java:94]: It shouldn't get to `new File(configuration.configurationFile())`, because `new File(configuration.configurationFile()).isAbsolute()` should be false.

BTW how are you testing this?

Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java:94]: in the playground project for now of tristan

File: server/testdriver/core/pom.xml (hunk 1 above)
Comment (karesti) [reply server/testdriver/core/pom.xml:84]: ok

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java (hunk 2 above)
Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java:46]: oki

File: commons-test/pom.xml (hunk 5 above)
Comment (karesti) [reply commons-test/pom.xml:34]: #8177 the JUnit 5 component upgrade is in another PR now

File: server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/InfinispanServerRuleBuilder.java (hunk 6 above)
Comment (karesti) [reply server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/InfinispanServerRuleBuilder.java:34]: oki

**Comment by tristantarrant** [comment]
I'm enjoying the commit messages :)

**Comment by karesti** [comment]
@tristantarrant I forgot this branch was actually in the PR

**Comment by karesti** [comment]
@tristantarrant so the extension test runs in intellij because it detects junit5. test passes. 
however, this does not work in command line, because something is wrong  somewhere and I'm unable to make  work a junit5 test. I spent so much time yesterday, but unable to make it work. 
I did quit a lot refactoring to be able to reuse the code. Unlike rules, we only need an extension.
CI will tell if something is wrong in the refactoring for the exsisting tests. 
There is still room for cleanup and improvement of course, this is a preview PR.

**Comment by karesti** [comment]
@tristantarrant @danberindei I forgot to  push a needed change in the pom.xml Coming

**Comment by danberindei** [comment]
@karesti I have made the JUnit 5 test driver run the tests with JUnit 5 here: https://github.com/infinispan/infinispan/commit/8dde3ef4e02f5a032f5168d639f4f3345dca16e5

Somehow neither the indications in the JUnit 5 documentation nor the indications in the Surefire documentation worked, I had to attach a debugger to Surefire in order to figure out which dependencies I had to add and where.

* Copy the maven-surefire-plugin from one of the JUnit modules
* Replace provider surefire-junit47 with surefire-junit-platform
* Add junit-jupiter-engine dependency
* Set scope of JUnit dependencies in parent POM to test
* Remove TestNG dependency
* Remove Infinispan core compile dependency in server parent

**Comment by karesti** [comment]
> @karesti I have made the JUnit 5 test driver run the tests with JUnit 5 here: [8dde3ef](https://github.com/infinispan/infinispan/commit/8dde3ef4e02f5a032f5168d639f4f3345dca16e5)
> 
> Somehow neither the indications in the JUnit 5 documentation nor the indications in the Surefire documentation worked, I had to attach a debugger to Surefire in order to figure out which dependencies I had to add and where.
> 
> * Copy the maven-surefire-plugin from one of the JUnit modules
> * Replace provider surefire-junit47 with surefire-junit-platform
> * Add junit-jupiter-engine dependency
> * Set scope of JUnit dependencies in parent POM to test
> * Remove TestNG dependency
> * Remove Infinispan core compile dependency in server parent

I feel relief concerning the fact that nothing that was documented worked, because I really spent an enormous amount of time. You are the best DAAAN!!!!

**Comment by karesti** [comment]
@danberindei the test has run 
`Running org.infinispan.server.test.junit5.InfinispanServerExtensionTest
[2020-04-24T17:26:51.000Z] 13:26:08.880 [INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 8.446 s - in org.infinispan.server.test.junit5.InfinispanServerExtensionTest
[2020-04-24T17:26:51.000Z] 13:26:09.222 [INFO] 
[2020-04-24T17:26:51.000Z] 13:26:09.222 [INFO] Results:
[2020-04-24T17:26:51.000Z] 13:26:09.223 [INFO] 
[2020-04-24T17:26:51.000Z] 13:26:09.223 [INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0`

**Comment by karesti** [comment]
@danberindei do you think we need to pass to every subproject now?

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/HotRod.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+      return this;
+   }
+
+   public <K, V> RemoteCache<K, V> get() {
+      RemoteCacheManager remoteCacheManager = testClient.registerResource(testServer.newHotRodClient(clientConfiguration));
+      String name = testClient.getMethodName(qualifier);
+      return remoteCacheManager.getCache(name);
+   }
+
+   public <K, V> RemoteCache<K, V> create() {
```
Comment (danberindei) [inline server/testdriver/core/src/main/java/org/infinispan/server/test/core/HotRod.java:32]: `create` also seems ambiguous to me, it doesn't say it's creating the cache on the server, so it could be interpreted as "create the `RemoteCache` instance so that `get()` can use it".

File: server/testdriver/junit4/pom.xml [hunk 8]
```diff
@@ -49,6 +48,16 @@
...
       </dependency>
+      <dependency>
+         <groupId>junit</groupId>
+         <artifactId>junit</artifactId>
+         <scope>compile</scope>
+      </dependency>
+      <dependency>
+         <groupId>org.testng</groupId>
+         <artifactId>testng</artifactId>
+         <scope>test</scope>
```
Comment (danberindei) [inline server/testdriver/junit4/pom.xml:59]: Do we really need TestNG here? Maybe add a source comment to explain why

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractServerConfigBuilder.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+import java.util.Properties;
+
+/**
+ * Common code for JUnit 4 and Junit 5 Extension
+ *
+ * @param <T> type of builder
+ * @author Katia Aresti
+ * @since 11
+ */
+public abstract class AbstractServerConfigBuilder<T> {
```
Comment (danberindei) [inline server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractServerConfigBuilder.java:14]: I suggest `AbstractTestServerBuilder`, because it creates a `TestServer` instance. Although `TestServer` isn't just one server either, it's a cluster, so maybe `AbstractTestClusterBuilder`?

File: server/testdriver/junit4/src/test/java/InfinispanRuleBasicTest.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+   @Rule
+   public InfinispanServerTestMethodRule SERVER_TEST = new InfinispanServerTestMethodRule(SERVER);
+
+   @Test
+   public void testSingleServer() {
+      ConfigurationBuilder builder = new ConfigurationBuilder();
+      builder.clientIntelligence(ClientIntelligence.BASIC);
+      RemoteCache<String, String> cache = SERVER_TEST.hotrod().withClientConfiguration(builder).withCacheMode(CacheMode.DIST_SYNC).create();
+            cache.put("k1", "v1");
```
Comment (danberindei) [inline server/testdriver/junit4/src/test/java/InfinispanRuleBasicTest.java:26]: Wrong indent, and the previous line should be wrapped

File: server/testdriver/junit5/src/main/java/org/infinispan/server/test/junit5/InfinispanServerExtension.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.BeforeTestExecutionCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+
+public class InfinispanServerExtension implements
```
Comment (danberindei) [inline server/testdriver/junit5/src/main/java/org/infinispan/server/test/junit5/InfinispanServerExtension.java:17]: Missing javadoc

File: server/testdriver/junit5/src/main/java/org/infinispan/server/test/junit5/InfispanServerExtensionBuilder.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+   /**
+    * Use this method to instantiate a single clustered embedded server
+    *
+    * @return InfinispanServerExtension
+    */
+   public static InfinispanServerExtension server() {
+      return server(true);
+   }
+
+   public static InfinispanServerExtension server(boolean container) {
```
Comment (danberindei) [inline server/testdriver/junit5/src/main/java/org/infinispan/server/test/junit5/InfispanServerExtensionBuilder.java:24]: Is the `container` parameter still needed?

We also need a test to make sure it works, there is no usage ATM.

File: server/testdriver/junit5/src/main/java/org/infinispan/server/test/junit5/InfispanServerExtensionBuilder.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
+package org.infinispan.server.test.junit5;
+
+import org.infinispan.server.test.core.AbstractServerConfigBuilder;
+import org.infinispan.server.test.core.ServerRunMode;
+
+import static org.infinispan.server.test.core.AbstractInfinispanServerDriver.DEFAULT_CLUSTERED_INFINISPAN_CONFIG_FILE_NAME;
+
+/**
+ * Infinispan Server Extension Builder
```
Comment (danberindei) [inline server/testdriver/junit5/src/main/java/org/infinispan/server/test/junit5/InfispanServerExtensionBuilder.java:9]: Shouldn't we have more info about how to use the extension, or at least a `@see InfinispanServerExtension`?

File: server/testdriver/junit5/src/test/java/org/infinispan/server/test/junit5/InfinispanServerExtensionTest.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+         .runMode(ServerRunMode.EMBEDDED)
+         .build();
+
+   @Test
+   public void testSingleServer() {
+      ConfigurationBuilder builder = new ConfigurationBuilder();
+      builder.clientIntelligence(ClientIntelligence.BASIC);
+      TestClient client = SERVER.client();
+      RemoteCache<String, String> cache = client.hotrod().withClientConfiguration(builder).withCacheMode(CacheMode.DIST_SYNC).create();
+            cache.put("k1", "v1");
```
Comment (danberindei) [inline server/testdriver/junit5/src/test/java/org/infinispan/server/test/junit5/InfinispanServerExtensionTest.java:28]: Wrong indent

File: server/testdriver/junit5/src/main/java/org/infinispan/server/test/junit5/InfispanServerExtensionBuilder.java (hunk 13 above)
Comment (karesti) [reply server/testdriver/junit5/src/main/java/org/infinispan/server/test/junit5/InfispanServerExtensionBuilder.java:9]: Dan,  possibly, but not in this PR please. I can create a JIRA to document and improve after this gets in.

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/Base.java [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+import org.infinispan.commons.configuration.XMLStringConfiguration;
+import org.infinispan.configuration.cache.CacheMode;
+
+import java.util.EnumSet;
+
+/**
+ * @since 10
+ * @param <S>
+ */
+public abstract class Base<S extends Base<S>> implements Self<S> {
```
[the remark this answers is not in the document]
Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/Base.java:15]: These classes, Base, HotRod and Rest classes already existed, as inner classes inside the Junit 4 method Rule. I just moved them out to make them reusable across junit 4 and junit 5.
But I'm going to refactor the names in an additional commit now

File: server/testdriver/junit5/src/main/java/org/infinispan/server/test/junit5/InfinispanServerExtension.java (hunk 11 above)
Comment (karesti) [reply server/testdriver/junit5/src/main/java/org/infinispan/server/test/junit5/InfinispanServerExtension.java:17]: ok

File: server/testdriver/junit5/src/main/java/org/infinispan/server/test/junit5/InfispanServerExtensionBuilder.java (hunk 13 above)
Comment (karesti) [reply server/testdriver/junit5/src/main/java/org/infinispan/server/test/junit5/InfispanServerExtensionBuilder.java:9]: ok for the javadoc in the other file

File: server/testdriver/junit5/src/test/java/org/infinispan/server/test/junit5/InfinispanServerExtensionTest.java (hunk 14 above)
Comment (karesti) [reply server/testdriver/junit5/src/test/java/org/infinispan/server/test/junit5/InfinispanServerExtensionTest.java:28]: oki

File: server/testdriver/junit4/src/test/java/InfinispanRuleBasicTest.java (hunk 10 above)
Comment (karesti) [reply server/testdriver/junit4/src/test/java/InfinispanRuleBasicTest.java:26]: oki

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/Base.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+import org.infinispan.commons.api.CacheContainerAdmin;
+import org.infinispan.commons.configuration.BasicConfiguration;
+import org.infinispan.commons.configuration.Self;
+import org.infinispan.commons.configuration.XMLStringConfiguration;
+import org.infinispan.configuration.cache.CacheMode;
+
+import java.util.EnumSet;
+
+/**
+ * @since 10
```
[the remark this answers is not in the document]
Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/Base.java:12]: the version is ok
the author is tristan and description was missing. this already existed Dan. but I'm going to improve it in the next commit

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/ContainerInfinispanServerDriver.java [hunk 17]
```diff
@@ -218,6 +218,13 @@ protected void start(String name, File rootDir, String configurationFile) {
       }
    }
 
+   public InfinispanGenericContainer getSingleContainer(int i) {
```
[the remark this answers is not in the document]
Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/ContainerInfinispanServerDriver.java:221]: ok

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/HotRod.java (hunk 7 above)
Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/HotRod.java:32]: again, all this already existed

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/TestClient.java [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+         resources.forEach(Util::close);
+         resources.clear();
+      }
+   }
+
+   public void initResources() {
+      resources = new ArrayList<>();
+   }
+
+   public String addScript(RemoteCacheManager remoteCacheManager, String script) {
```
[the remark this answers is not in the document]
Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/TestClient.java:73]: possibly, yes. I kept things as untouched as possible in the test suite.
Again, this classes and methods already existed
I will try to refactor in another commit

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/TestClient.java [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+         throw new RuntimeException(e);
+      }
+      return getMethodName();
+   }
+
+   public String getMethodName() {
+      return getMethodName(null);
+   }
+
+   public RestClient newRestClient(RestClientConfigurationBuilder restClientConfigurationBuilder) {
```
[the remark this answers is not in the document]
Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/TestClient.java:87]: ok

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/TestClient.java [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+         MessageDigest sha1 = MessageDigest.getInstance("SHA-1");
+         byte[] digest = sha1.digest(cacheName.getBytes(StandardCharsets.UTF_8));
+         return Util.toHexString(digest);
+      } catch (NoSuchAlgorithmException e) {
+         // Won't happen
+         return null;
+      }
+   }
+
+   public MemcachedClient getMemcachedClient() {
```
[the remark this answers is not in the document]
Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/TestClient.java:105]: again, this is the testsuite impact. I will refactor then

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/TestServer.java [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+import org.testcontainers.DockerClientFactory;
+
+import java.io.Closeable;
+import java.io.File;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+
+public class TestServer {
```
[the remark this answers is not in the document]
Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/TestServer.java:18]: ok

File: server/testdriver/junit5/src/main/java/org/infinispan/server/test/junit5/InfinispanServerExtension.java (hunk 11 above)
Comment (karesti) [reply server/testdriver/junit5/src/main/java/org/infinispan/server/test/junit5/InfinispanServerExtension.java:17]: ok

File: server/testdriver/junit5/src/main/java/org/infinispan/server/test/junit5/InfispanServerExtensionBuilder.java (hunk 12 above)
Comment (karesti) [reply server/testdriver/junit5/src/main/java/org/infinispan/server/test/junit5/InfispanServerExtensionBuilder.java:24]: adding a test

File: server/testdriver/junit4/pom.xml (hunk 8 above)
Comment (karesti) [reply server/testdriver/junit4/pom.xml:59]: checking

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/TestClient.java [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+
+import java.io.Closeable;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class TestClient {
```
[the remark this answers is not in the document]
Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/TestClient.java:22]: ok

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/InfinispanServerTestConfiguration.java [hunk 23]
```diff
@@ -18,10 +18,11 @@
    private final JavaArchive[] archives;
    private final boolean jmx;
    private final boolean parallelStartup;
+   private boolean defaultFile;
 
    public InfinispanServerTestConfiguration(String configurationFile, int numServers, ServerRunMode runMode,
                                             Properties properties, String[] mavenArtifacts, JavaArchive[] archives,
-                                            boolean jmx, boolean parallelStartup) {
+                                            boolean jmx, boolean parallelStartup, boolean defaultFile) {
```
[the remark this answers is not in the document]
Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/InfinispanServerTestConfiguration.java:25]: yes. in container mode. we don't copy anything to the container run that way

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/HotRod.java [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+      this.testServer = testServer;
+      this.testClient = testClient;
+   }
+
+   public HotRod withClientConfiguration(ConfigurationBuilder clientConfiguration) {
+      this.clientConfiguration = clientConfiguration;
+      return this;
+   }
+
+   public <K, V> RemoteCache<K, V> get() {
```
[the remark this answers is not in the document]
Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/HotRod.java:26]: I will review this, but this impacts all our test suite. I just moved code around. not sure I should change this in this PR though.

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractServerConfigBuilder.java (hunk 9 above)
Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractServerConfigBuilder.java:14]: I don't agree. The Rule is called "InfinispanServerRuleBuilder" not "InfinispanClusterRuleBuilder". This builder emerged up to reuse the same type of build for the InfinispanServerExtension and InfinispanServerRule.
The InfinispanServer is either a cluster or a standalone server.

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java [hunk 25]
```diff
@@ -39,6 +39,8 @@
  * @since 10.0
  **/
 public abstract class AbstractInfinispanServerDriver implements InfinispanServerDriver {
+   public static final String DEFAULT_CLUSTERED_INFINISPAN_CONFIG_FILE_NAME = "infinispan.xml";
```
[the remark this answers is not in the document]
Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java:42]: yes

File: server/testdriver/core/pom.xml [hunk 26]
```diff
@@ -67,6 +97,11 @@
          <groupId>net.spy</groupId>
          <artifactId>spymemcached</artifactId>
       </dependency>
+      <dependency>
+         <groupId>junit</groupId>
+         <artifactId>junit</artifactId>
+         <scope>compile</scope>
```
[the remark this answers is not in the document]
Comment (karesti) [reply server/testdriver/core/pom.xml:73]: yes

File: build-configuration/pom.xml [hunk 27]
```diff
@@ -172,7 +172,8 @@
       <version.jetty>9.4.5.v20170502</version.jetty>
       <version.jgroups>4.2.1.Final</version.jgroups>
       <version.jsr107>1.1.0</version.jsr107>
-      <version.junit>4.13</version.junit>
+      <version.junit4>4.13</version.junit4>
+      <version.junit>5.6.2</version.junit>
```
[the remark this answers is not in the document]
Comment (karesti) [reply build-configuration/pom.xml]: ok

**Comment by karesti** [comment]
@danberindei I rebased and created a commit with your latest review

**Comment by karesti** [comment]
> I'm not convinced we need separate `TestServer` and `TestClient` classes, I'd go with a single `TestCluster` or `TestServerCluster`.

I will review this part tomorrow and add another commit with the changes

**Comment by karesti** [comment]
> Needs squashing and some more javadocs.
> 
> I'm not convinced we need separate `TestServer` and `TestClient` classes, I'd go with a single `TestCluster` or `TestServerCluster`.
> 
> The `Base`, `HotRod`, and `Rest` class names probably worked ok when they were inner classes, but as top-level classes they seem too short. IMO their methods could be moved in the same `Test[Server]Cluster` class without losing anything.

Well the thing is that the code in testserver corresponds to the code that is used in the class rule that creates the server, and the test client is used in each method. I took the code and separated in two classes, because this was already done this way between the server rule and the method rule.

And in the server extension I do the same thing. The difference is we don't need two extensions, but a single one, and deal in the methods what corresponds to the server rule and the method rule.

For me in this case you are nitpicking with an implementation detail that was already there, that does not affect at all to the end user testing, because it is internal.

However, concerning the HotRod Rest or other, we can find another name if you want. Or try to decide on a end-user API. We have time for that now still, since this code is still only used in our code base

Could we rethink this, please, trying to have this PR in the next DEV release and feel how it looks like to use the Junit 5 extension, with an actual end user testing it ? And interate from there ?
I'm afraid to stuck too long here and not getting things done.

**Comment by danberindei** [comment]
@tristantarrant please have a look at the API as well

**Comment by tristantarrant** [comment]
@karesti the PR looks good now, aside from some conflicts introduced by my changed related to the database tests.
Fix those and we can merge it.
We then need three additional steps:
* documentation on how a user would use this in her own tests (@oraNod I guess we need a new "testing" book)
* migrating our server test suite to use this
* migrate my https://github.com/tristantarrant/infinispan-playground-testdriver to use this and bring it to the infinispan organization.

**Comment by karesti** [comment]
@tristantarrant rebased. let's see how CI behaves

**Comment by karesti** [comment]
> * migrate my https://github.com/tristantarrant/infinispan-playground-testdriver to use this and bring it to the infinispan organization.

Since I agree about the 3 points, I wouldn't bring the playground to the organisation.
Instead, add simple tutorials showcasing several use cases and use it to cover unit testing for remote examples.

**Comment by karesti** [comment]
Since the server is auth by default now, I'm adding a commit to provide default auth config

commit coming soon

**Comment by karesti** [comment]
there are related tests

File: server/testdriver/core/src/main/java/org/infinispan/server/test/api/HotRodTestClientDriver.java [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+   private final TestServer testServer;
+   private final TestClient testClient;
+   ConfigurationBuilder clientConfiguration;
+
+   public HotRodTestClientDriver(TestServer testServer, TestClient testClient) {
+      this.testServer = testServer;
+      this.testClient = testClient;
+
+      ConfigurationBuilder builder = new ConfigurationBuilder();
+      if(testServer.isContainerRunWithDefaultServerConfig()) {
```
Comment (danberindei) [inline server/testdriver/core/src/main/java/org/infinispan/server/test/api/HotRodTestClientDriver.java:27]: AFAICT `AbstractInfinispanServerDriver.createUserFile` runs with a non-default configuration just as well as with the default configuration, so this check isn't necessary.

Comment (karesti) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/api/HotRodTestClientDriver.java:27]: it is necessary since I only want to add this configuration for container mode and when the default file is requested (which is secured). Otherwise, users need to concern too early about security. But I don't want to make this in embedded since default config files stuff are not supported yet, and there are XML files that are not secured in our tests. So if I add a user admin, I get hard error because the server is not secured in the test but I try to connect with a user. (check the other runs errors)
I don't know how to make this better for now.

**Comment by tristantarrant** [comment]
Merged. Thanks a lot Katia

**Comment by karesti** [comment]
@tristantarrant I'm hearing angels singing

File: server/testdriver/core/src/main/java/org/infinispan/server/test/api/HotRodTestClientDriver.java (hunk 28 above)
Comment (danberindei) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/api/HotRodTestClientDriver.java:27]: Thanks for the explanation Katia, I thought that the users were added for every configuration.
